spark-unified/src/main/scala/io/delta/sql/ApplyV2Streaming.scala

```diff
@@ -0,0 +1,135 @@
+import org.apache.spark.sql.connector.catalog.Identifier
+import org.apache.spark.sql.execution.streaming.StreamingRelation
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+/**
+ * Rule for applying the V2 streaming path by rewriting V1 StreamingRelation
+ * with Delta DataSource to StreamingRelationV2 with SparkTable.
+ *
+ * This rule handles the case where Spark's FindDataSourceTable rule has converted
+ * a StreamingRelationV2 (with DeltaTableV2) back to a StreamingRelation because
+ * DeltaTableV2 doesn't advertise STREAMING_READ capability. We convert it back to
+ * StreamingRelationV2 with SparkTable (from kernel-spark) which does support streaming.
+ *
+ * Behavior based on spark.databricks.delta.v2.enableMode:
+ * - AUTO (default): Only applies to Unity Catalog managed tables
+ * - STRICT: Applies to all Delta tables (for testing V2 streaming)
+ * - NONE: Rule is disabled, no conversion happens
+ *
+ * @param session The Spark session for configuration access
+ */
+class ApplyV2Streaming(
+    @transient private val session: SparkSession)
+  extends Rule[LogicalPlan] {
+
+  private def isDeltaStreamingRelation(s: StreamingRelation): Boolean = {
+    // Check if this is a Delta streaming relation by examining:
+    // 1. The source name (e.g., "delta" from .format("delta"))
+    // 2. The catalog table's provider (e.g., "DELTA" from Unity Catalog)
+    // 3. Whether the table is a Unity Catalog managed table
+    s.dataSource.catalogTable match {
+      case Some(catalogTable) =>
+        DeltaSourceUtils.isDeltaDataSourceName(s.sourceName) ||
+        catalogTable.provider.exists(DeltaSourceUtils.isDeltaDataSourceName) ||
+        CatalogTableUtils.isUnityCatalogManagedTable(catalogTable)
+      case None => false
+    }
+  }
+
+  private def shouldApplyV2Streaming(s: StreamingRelation): Boolean = {
+    if (!isDeltaStreamingRelation(s)) {
+      return false
+    }
+
+    val mode = session.conf.get(
+      DeltaSQLConfV2.V2_ENABLE_MODE.key,
+      DeltaSQLConfV2.V2_ENABLE_MODE.defaultValueString)
+
+    // scalastyle:off caselocale
+    mode.toUpperCase match {
+    // scalastyle:on caselocale
+      case "STRICT" =>
+        // Always apply V2 streaming for all Delta tables
+        true
+      case "AUTO" =>
+        // Only apply for Unity Catalog managed tables
+        // catalogTable is guaranteed to be Some because isDeltaStreamingRelation checked it
+        s.dataSource.catalogTable.exists(CatalogTableUtils.isUnityCatalogManagedTable)
+      case "NONE" | _ =>
+        // V2 streaming disabled
+        false
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+    case s: StreamingRelation if shouldApplyV2Streaming(s) =>
+      // catalogTable is guaranteed to be defined because shouldApplyV2Streaming checks it
+      // via isDeltaStreamingRelation, but we use pattern matching for safety
+      s.dataSource.catalogTable match {
```

Comment:
if `shouldApplyV2Streaming` returns yes, `s.dataSource.catalogTable` should be defined